src/test/script_P2CS_tests.cpp

```diff
@@ -197,4 +199,105 @@ BOOST_AUTO_TEST_CASE(coldstake_script)
     BOOST_CHECK(CheckP2CSScript(tx.vin[0].scriptSig, scriptP2CS, tx, err));
 }
 
+// Check that it's not possible to "fake" a P2CS script for the owner by splitting the locking
+// and unlocking parts. This particular script can be spent by any key, with a
+// unlocking script composed like: <sig> <pk> <DUP> <HASH160> <pkh>
+static CScript GetFakeLockingScript(const CKeyID staker, const CKeyID& owner)
+{
+    CScript script;
+    script << opcodetype(0x2F) << opcodetype(0x01) << OP_ROT <<
+            OP_IF << OP_CHECKCOLDSTAKEVERIFY << ToByteVector(staker) <<
+            OP_ELSE << ToByteVector(owner) << OP_DROP <<
+            OP_EQUALVERIFY << OP_CHECKSIG;
+
+    return script;
+}
+
+void FakeUnlockColdStake(CMutableTransaction& tx, const CScript& prevScript, const CKey& key)
+{
+    // sign the first input
+    tx.vin[0].scriptSig.clear();
+    const CTransaction _tx(tx);
+    SigVersion sv = _tx.GetRequiredSigVersion();
+    const uint256& hash = SignatureHash(prevScript, _tx, 0, SIGHASH_ALL, amtIn, sv);
+    std::vector<unsigned char> vchSig;
+    BOOST_CHECK(key.Sign(hash, vchSig));
+    vchSig.push_back((unsigned char)SIGHASH_ALL);
+    tx.vin[0].scriptSig << vchSig << ToByteVector(key.GetPubKey()) << OP_DUP << OP_HASH160 << ToByteVector(key.GetPubKey().GetID());
+}
+
+static void setupWallet(CWallet& wallet)
+{
+    wallet.SetMinVersion(FEATURE_SAPLING);
+    wallet.SetupSPKM(false);
+}
+
+BOOST_AUTO_TEST_CASE(fake_script_test)
+{
+    BOOST_ASSERT(!g_IsV6Active);
+
+    CWallet& wallet = *pwalletMain;
+    LOCK(wallet.cs_wallet);
+    setupWallet(wallet);
+    CKey stakerKey;         // dummy staker key (not in the wallet)
+    stakerKey.MakeNewKey(true);
+    CKeyID stakerId = stakerKey.GetPubKey().GetID();
+    CPubKey ownerPubKey;
+    BOOST_ASSERT(wallet.GetKeyFromPool(ownerPubKey));
+    const CKeyID& ownerId = ownerPubKey.GetID();
+    CKey ownerKey;          // owner key (in the wallet)
+    BOOST_ASSERT(wallet.GetKey(ownerId, ownerKey));
+
+    const CScript& scriptP2CS = GetFakeLockingScript(stakerId, ownerId);
+
+    // Create prev transaction:
+    // It has two outputs. The first one is spent before v6.
+    // The second one is tested after v6 enforcement.
+    CMutableTransaction txFrom;
+    txFrom.vout.resize(2);
```

Comment:
No need to create two outputs.
Plus, the post v6 enforcement check that is mentioned in the comment is not being done (you could do it with the first output as well. No difference between them at the interpreter-stack level. The script will pass the verification as well, only difference is that the wallet will handle it correctly after v6)

Reply:
We only need to check that, after v6, the output is no longer credited to the owner's wallet. That's what the test is doing.
We don't need to re-test the interpreter, as nothing changed there.

Reply:
yeah, was pointing that. But nvm, dunno why i thought that you there with the  "tested after v6" line were referring to the script validation.